File: Java/Q206ReverseLinkedList.java
/*
    @b-knd (jingru) on 31 July 2022 10:10:00
    
    Concept:
    - having three pointer: curr, prev and next
    - basically change next pointer to make it point to previous node (prev pointer)
    - prev pointer always points to the last modified node (beginning of the reversed part of the list)
    - then advance curr to the next node in original list
*/

class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode curr = head, prev = null, next = null;
        
        while(curr != null){
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        
        return prev;
    }
    
    //Runtime: 0 ms, faster than 100.00% of Java online submissions for Reverse Linked List.
    //Memory Usage: 43.2 MB, less than 27.01% of Java online submissions for Reverse Linked List.
}

File: Python/Q206ReverseLinkedList.py
# @b-knd (jingru) on 10 August 2022 08:41:00

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        nextNode = None
        
        while curr != None:        
            nextNode = curr.next
            curr.next = prev
            prev = curr
            curr = nextNode
        
        return prev

#Runtime: 41 ms, faster than 86.59% of Python3 online submissions for Reverse Linked List.
#Memory Usage: 15.4 MB, less than 94.21% of Python3 online submissions for Reverse Linked List.
